### ISSUE TITLE
Server connection was stucked (probable regression in 6.6.4 since 6.4.2)

### ISSUE THREAD

**Issue by cdbennett** [mosquito/aiormq#165]
I'm using aio-pika with aiormq. Until I updated to the latest releases of these components, the reconnection support was working great.

Before (working fine):  aio-pika 8.2.4   aiormq 6.4.2
Now (stucked):  aio-pika 8.3.0  aiormq 6.6.4

When the Python application apparently had lost its connection to the RabbitMQ server and tried to reconnect, it got into a state where it would never reconnect. Stack traces were printed.

```
2023-02-01 04:37:32,928 INFO [aio_pika.robust_connection] Connection to amqp://guest:******@localhost:8121//?fail_fast=0 closed. Reconnecting after 5 seconds.
2023-02-01 08:25:01,776 ERROR [aio_pika.robust_connection] Failed to reopen channel
Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 179, in rpc
    result = await countdown(self.rpc_frames.get())
  File "/app/env/lib/python3.10/site-packages/aiormq/tools.py", line 95, in __call__
    return await coro
  File "/app/env/lib/python3.10/asyncio/queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 86, in __connection_attempt
    await channel.restore(connection)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 79, in restore
    await self.reopen()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 202, in reopen
    await self._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 82, in _open
    await super()._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 151, in _open
    channel = await UnderlayChannel.create(
  File "/app/env/lib/python3.10/site-packages/aio_pika/abc.py", line 451, in create
    channel = await connection.channel(**kwargs)
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 783, in channel
    await channel.open(timeout=timeout)
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 216, in open
    frame: spec.Channel.OpenOk = await self.rpc(
  File "/app/env/lib/python3.10/site-packages/aiormq/base.py", line 164, in wrap
    return await self.create_task(func(self, *args, **kwargs))
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 590, in __reader
    raise asyncio.TimeoutError(
asyncio.exceptions.TimeoutError: Server connection was stucked. No frames were received in 183 seconds.
```

The RabbitMQ server did not seem to the cause:
- RabbitMQ server log didn't have any unusual log output
- Other clients could connect to RabbitMQ server fine
- Restarting the RabbitMQ server didn't fix it

After hitting Ctrl+c on the Python app, it printed more exceptions on shutdown:

```
2023-02-01 11:32:43,416 DEBUG [main] Stopped RabbitMQ task
2023-02-01 11:32:43,419 INFO [aio_pika.queue] <RobustQueueIterator: queue='request' ctag='ctag1.8a18ba44854b417d9d93d8236da113bb'> closing with timeout 5 seconds
2023-02-01 11:32:43,847 ERROR [asyncio] Task exception was never retrieved
future: <Task finished name='Task-2088' coro=<OneShotCallback.__task_inner() done, defined at /app/env/lib/python3.10/site-packages/aio_pika/tools.py:230> exception=TimeoutError('Server connection was stucked. No frames were received in 183 seconds.')>
Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 179, in rpc
    result = await countdown(self.rpc_frames.get())
  File "/app/env/lib/python3.10/site-packages/aiormq/tools.py", line 95, in __call__
    return await coro
  File "/app/env/lib/python3.10/asyncio/queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aio_pika/tools.py", line 236, in __task_inner
    await self.callback(*args, **kwargs)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 73, in _on_connection_close
    await self.reconnect()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 157, in reconnect
    await self.connect()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 124, in connect
    connection = await self.__connection_attempt(timeout)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 86, in __connection_attempt
    await channel.restore(connection)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 79, in restore
    await self.reopen()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 202, in reopen
    await self._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 82, in _open
    await super()._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 151, in _open
    channel = await UnderlayChannel.create(
  File "/app/env/lib/python3.10/site-packages/aio_pika/abc.py", line 451, in create
    channel = await connection.channel(**kwargs)
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 783, in channel
    await channel.open(timeout=timeout)
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 216, in open
    frame: spec.Channel.OpenOk = await self.rpc(
  File "/app/env/lib/python3.10/site-packages/aiormq/base.py", line 164, in wrap
    return await self.create_task(func(self, *args, **kwargs))
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 590, in __reader
    raise asyncio.TimeoutError(
asyncio.exceptions.TimeoutError: Server connection was stucked. No frames were received in 183 seconds.
2023-02-01 11:32:43,854 ERROR [asyncio] Future exception was never retrieved
future: <Future finished exception=TimeoutError('Server connection was stucked. No frames were received in 183 seconds.')>
Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 179, in rpc
    result = await countdown(self.rpc_frames.get())
  File "/app/env/lib/python3.10/site-packages/aiormq/tools.py", line 95, in __call__
    return await coro
  File "/app/env/lib/python3.10/asyncio/queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/env/lib/python3.10/site-packages/aio_pika/tools.py", line 236, in __task_inner
    await self.callback(*args, **kwargs)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 73, in _on_connection_close
    await self.reconnect()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 157, in reconnect
    await self.connect()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 124, in connect
    connection = await self.__connection_attempt(timeout)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_connection.py", line 86, in __connection_attempt
    await channel.restore(connection)
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 79, in restore
    await self.reopen()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 202, in reopen
    await self._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/robust_channel.py", line 82, in _open
    await super()._open()
  File "/app/env/lib/python3.10/site-packages/aio_pika/channel.py", line 151, in _open
    channel = await UnderlayChannel.create(
  File "/app/env/lib/python3.10/site-packages/aio_pika/abc.py", line 451, in create
    channel = await connection.channel(**kwargs)
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 783, in channel
    await channel.open(timeout=timeout)
  File "/app/env/lib/python3.10/site-packages/aiormq/channel.py", line 216, in open
    frame: spec.Channel.OpenOk = await self.rpc(
  File "/app/env/lib/python3.10/site-packages/aiormq/base.py", line 164, in wrap
    return await self.create_task(func(self, *args, **kwargs))
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 42, in __inner
    raise self.exception from e
  File "/app/env/lib/python3.10/asyncio/tasks.py", line 650, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/app/env/lib/python3.10/site-packages/aiormq/abc.py", line 40, in __inner
    return await self.task
  File "/app/env/lib/python3.10/site-packages/aiormq/connection.py", line 590, in __reader
    raise asyncio.TimeoutError(
asyncio.exceptions.TimeoutError: Server connection was stucked. No frames were received in 183 seconds.
```

Seems likely to be a bug in aiormq.
Here's a commit that recently (aiormq >= 6.5.0) touched the connection.py file and the lines of code related to the "connection was stucked" log message: https://github.com/mosquito/aiormq/commit/b209afd036787e3b4d5605beec47a867bee341d2

**Comment by mosquito** [comment]
Should be fixed in `aiormq>=6.7.0`

**Comment by lpdswing** [comment]
i have same issue ,aio-pika==9.2.2,aiormq ==6.7.7

**Comment by mosquito** [comment]
@lpdswing, please point out how I can reproduce this.

**Comment by lpdswing** [comment]
> 
I use the aio_pika I use aio pika as the broker of taskiq.
If there is no operation for a period of time, it is found that there is a warning that the amqp connection was stuck.
the warning log:
`aiormq.connection:__reader:668 - Server connection <Connection: "amqp://guest:******@127.0.0.1:5672/" at 0x1152a6ee0> was stuck. No frames were received in 183 seconds.`
the rabbitmq log:
```
tmp-rmq-1  | 2023-08-23 05:52:17.781108+00:00 [info] <0.15691.1> accepting AMQP connection <0.15691.1> (172.24.0.1:36618 -> 172.24.0.3:5672)                │tmp-kafka-1  | [2023-08-23 02:47:54,963] INFO [SnapshotEmitter id=0] Successfully
tmp-rmq-1  | 2023-08-23 05:52:18.335742+00:00 [info] <0.15691.1> connection <0.15691.1> (172.24.0.1:36618 -> 172.24.0.3:5672): user 'guest' authenticated an│ wrote snapshot 00000000000000100536-0000000001 (org.apache.kafka.image.publisher
d granted access to vhost '/'                                                                                                                               │.SnapshotEmitter)
tmp-rmq-1  | 2023-08-23 05:56:25.725122+00:00 [error] <0.15486.1> closing AMQP connection <0.15486.1> (172.24.0.1:32788 -> 172.24.0.3:5672):                │tmp-kafka-1  | [2023-08-23 03:47:50,990] INFO [SnapshotGenerator id=0] Creating n
tmp-rmq-1  | 2023-08-23 05:56:25.725122+00:00 [error] <0.15486.1> missed heartbeats from client, timeout: 60s                                               │ew KRaft snapshot file snapshot 00000000000000107728-0000000001 because we have w
tmp-rmq-1  | 2023-08-23 05:57:14.746616+00:00 [error] <0.15553.1> closing AMQP connection <0.15553.1> (172.24.0.1:46128 -> 172.24.0.3:5672):                │aited at least 60 minute(s). (org.apache.kafka.image.publisher.SnapshotGenerator)
tmp-rmq-1  | 2023-08-23 05:57:14.746616+00:00 [error] <0.15553.1> missed heartbeats from client, timeout: 60s                                               │tmp-kafka-1  | [2023-08-23 03:47:51,015] INFO [SnapshotEmitter id=0] Successfully
tmp-rmq-1  | 2023-08-23 05:58:02.029089+00:00 [error] <0.15624.1> closing AMQP connection <0.15624.1> (172.24.0.1:42672 -> 172.24.0.3:5672):                │ wrote snapshot 00000000000000107728-0000000001 (org.apache.kafka.image.publisher
tmp-rmq-1  | 2023-08-23 05:58:02.029089+00:00 [error] <0.15624.1> missed heartbeats from client, timeout: 60s
```
the rabbitmq docker-compose file:
```yml
version: "3.9"
services:
  rmq:
    image: rabbitmq:management
    ports:
      - '5672:5672'
      - '15672:15672'
    restart: always
    environment:
      RABBITMQ_DEFAULT_USER: "guest"
      RABBITMQ_DEFAULT_PASS: "guest"
      RABBITMQ_DEFAULT_VHOST: "/"
    healthcheck:
      test: rabbitmq-diagnostics check_running -q
      interval: 3s
      timeout: 3s
      retries: 50
```
in my project use:
https://github.com/lpdswing/fastapi-zero/blob/main/%7B%7Bcookiecutter.project_name%7D%7D/backend/src/lib/rabbitmq/lifespan.py

https://github.com/lpdswing/fastapi-zero/blob/main/%7B%7Bcookiecutter.project_name%7D%7D/backend/src/utils/router.py

**Comment by aozupek** [comment]
+1

**Comment by mosquito** [comment]
Thanks to @aozupek the problem is resolved and a new fixed version released as [aio-pika==9.2.3](https://pypi.org/project/aio-pika/9.2.3/)
